doc: highlight recommendation of eachindex in arrays.md (#45341)
JuliaLang/julia: doc/src/manual/arrays.md

are implemented using a linear chunk of memory and directly use a linear index
in their implementations, other arrays — like [`Diagonal`](@ref) — need the
full set of cartesian indices to do their lookup (see [`IndexStyle`](@ref) to

> introspect which is which). 

Reviewer: ```suggestion
introspect which is which).
```

!!! warnings

    When iterating over an entire array, it is

>     better to iterate over [`eachindex(A)`](@ref) instead of `1:length(A)`.

Reviewer 2: Not new to this PR, but "iterating over an ... array" seems a bit confusing, since you can `iterate` arrays but not like this.

Maybe it should say "iterating over all indices of an array". Maybe it should mention that you can iterate the array itself, which is probably better when you don't actually need the index.

>     When iterating over an entire array, it's
>     much better to iterate over [`eachindex(A)`](@ref) instead of `1:length(A)`.
>     Not only will the former be much faster in cases where `A` is `IndexCartesian`,
>     but it will also support [OffsetArrays](https://github.com/JuliaArrays/OffsetArrays.jl), too.

Reviewer: I'm not a native English speaker, but I'm not sure "much better" fits a formal text like documentation.  Also, I'd generalise the conclusion to not make it sound `OffsetArrays` is any special, but only one instance of arrays with custom indexing:
```suggestion
    When iterating over an entire array, it is
    better to iterate over [`eachindex(A)`](@ref) instead of `1:length(A)`.
    Not only will the former be much faster in cases where `A` is `IndexCartesian`,
    but it will also support arrays with custom indexing, such as [OffsetArrays](https://github.com/JuliaArrays/OffsetArrays.jl).
```

are implemented using a linear chunk of memory and directly use a linear index
in their implementations, other arrays — like [`Diagonal`](@ref) — need the
full set of cartesian indices to do their lookup (see [`IndexStyle`](@ref) to
introspect which is which).

!!! warnings

>     When iterating over the an entire array, and the values of the indices are needed, is
>     better to iterate over [`eachindex(A)`](@ref) instead of `1:length(A)`.

Reviewer 3: I could be wrong since I'm not a native English speaker. My limited English grammar knowledge tells me that it should be:

```suggestion
    When iterating over the entire array with the values of the indices needed, it is
    better to iterate over [`eachindex(A)`](@ref) instead of `1:length(A)`.
```

are implemented using a linear chunk of memory and directly use a linear index
in their implementations, other arrays — like [`Diagonal`](@ref) — need the
full set of cartesian indices to do their lookup (see [`IndexStyle`](@ref) to
introspect which is which).

!!! warnings

>     When iterating over the entire array with the values of the indices needed, it is
>     better to iterate over [`eachindex(A)`](@ref) instead of `1:length(A)`.
>     Not only will the former be faster in cases where `A` is `IndexCartesian`,
>     but it will also support arrays with custom indexing, such as [OffsetArrays](https://github.com/JuliaArrays/OffsetArrays.jl).
>     If the values of the indices are not needed, on the other hand, it is preferable to iterate over
>     the array's elements (i.e. `for a in A`).

Reviewer 2: Can this be shorter and clearer?  "values of the indices" seems a little odd, aren't those just indices... and values the things in the array?

```suggestion
    When iterating over all the indices for an array, it is
    better to iterate over [`eachindex(A)`](@ref) instead of `1:length(A)`.
    Not only will this be faster in cases where `A` is `IndexCartesian`,
    but it will also support arrays with custom indexing, such as [OffsetArrays](https://github.com/JuliaArrays/OffsetArrays.jl).
    If only the values are needed, then is better to just iterate the array directly, i.e. `for a in A`.
```

i = CartesianIndex(3, 2)
```

!!! note

>     In contrast with `for i = 1:length(A)`, iterating with [`eachindex`](@ref) provides an efficient way to
>     iterate over any array type.

Reviewer 3: ```suggestion
    In contrast with `for i = 1:length(A)`, iterating with [`eachindex`](@ref) provides an efficient way to
    iterate over any array type. Besides, this also supports generic arrays with custom indexing such as
    [OffsetArrays](https://github.com/JuliaArrays/OffsetArrays.jl).
```
